core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -516,6 +516,10 @@ object SparkSubmit extends CommandLineUtils {
     if (deployMode == CLIENT || isYarnCluster) {
       childMainClass = args.mainClass
       if (isUserJar(args.primaryResource)) {
+        val hadoopConf = new HadoopConfiguration()
+        args.primaryResource =
+          Option(args.primaryResource).map(
+            downloadFile(_, targetDir, args.sparkProperties, hadoopConf)).orNull
         childClasspath += args.primaryResource
       }
       if (args.jars != null) { childClasspath ++= args.jars.split(",") }
```

Comment:
I think in your scenario we should also download jars specified with `--jars` to local and add to classpath.